### ISSUE TITLE
Retrieve all route segments configuration by URL

### ISSUE THREAD

**Issue by josephperrott** [angular/angular#46381]
**Note**: This issue was recreated from the discussion as we over eagerly converted it to a discussion before we had confirmed our process for potentially using discussions for this type of issue.  I have moved it to back to an issue, and you can see the original issue with comments at https://github.com/angular/angular/issues/15750.

<div type='discussions-op-text'>

<sup>Originally posted by **smnbbrv** April  4, 2017</sup>

**I'm submitting a ...**  (check one with "x")
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

There is no clear way to get the configuration, see http://stackoverflow.com/q/43143621/1990451.

Given a URL, e.g. `/path/path2/path3/123;sdf=df` and a routes configuration:

    { path: 'path', data: { g: 'h' }, children: [
      { path: 'path2', data: { c: 'd' }, children: [
        { path: 'something', component: TestpageComponent, data: { a: 'b' } },
        { path: 'path3/create', component: TestpageComponent, data: { b: 'c' } },
        { path: 'path3/:id', component: TestpageComponent, data: { b: 'c' } },
      ] }
    ] },

It would be useful to find an actual configuration for all the segments in order to e.g. get a result set of all `data` parameters on all levels (each segment of the url).

It is possible to split the URL to the segments by e.g.

    console.log(this.router.parseUrl(url));

or to be more precise

    console.log(this.router.parseUrl(url).root.children['primary'].segments);

which will return 

    [{"path":"path","parameters":{}},{"path":"path2","parameters":{}},{"path":"and","parameters":{}},{"path":"123","parameters":{"sdf":"df"}}]

So, splitting the URL into segments is not a problem. However it does not provide the configuration for each segment. One can get the actual configuration for all the routes with

    console.log(this.router.config);

It's possible to go with the parsed segments through the configuration tree and resolve the branch (and it's not that complicated actually), but it can cause troubles e.g. while resolving `:id` against `create` segment. So, it would be cool to use the *router's* way of resolving the configuration.

**Expected behavior**

I would expect a method on a `Router` service that returns an array of all the configurations available for a given URL. E.g.

    router.configurationByUrl('/path/path2/path3/123;sdf=df')
    // returns [
    //  { path: 'path', data: { g: 'h' }, children: [...] },
    //  { path: 'path2', data: { c: 'd' }, children: [...] },
    //  { path: 'path3/:id', component: TestpageComponent, data: { b: 'c' } }
    //], 

**Minimal reproduction of the problem with instructions**

**What is the motivation / use case for changing the behavior?**

Imagine I have a 50% of the URLs protected by some guard (e.g. only logged in users could go there) and other 50% is not protected (displayed for everybody). So, on the navigation menu level (outside of the `router-outlet`) I want to display only the links which are relevant for the current user, so I need to know which routes are protected by the guard. Guards / data / whatever is just a particular case of the problem.

**Please tell us about your environment:**

* **Angular version:** 2.0.X / 4.0.X

* **Browser:** [all]

 
* **Language:** [all]

* **Node (for AoT issues):** `node --version` =  X</div>

**Comment by atscott** [comment]
An additional option to the [one which uses unsupported private APIs](https://github.com/angular/angular/issues/15750#issuecomment-292238096) would be to create a second `Router` detached from the application's injector tree and perform "dry run" navigations through it by providing a `PlatformLocation` which doesn't touch the browser history: https://stackblitz.com/edit/angular-ivy-z2xj2g?file=src%2Fapp%2Fapp.component.ts

I still wouldn't say this is a great idea but it's an interesting experiment. The listed motivation doesn't feel compelling enough to introduce such a complex solution. The location that lists the links could much more easily just decide on the correct conditions for showing the routes. That component's knowledge of the existence of the routes indicates it can also have knowledge of whether those routes should be accessible.

**Comment by AndreasHae** [comment]
YES! The specific use case I'd need this for: I have a route guard that denies access based on user permissions specified in the route data. Now I want to hide the menu bar items linking to pages the user doesn't have access to.

Currently I have the user permission config duplicated in the menu bar config and the router config, which is error-prone and tedious to maintain. I'd love to be able to extract the router config based on the router link the menu item leads to so that I have a single source of truth.
